src/basic/conf-files.c

```diff
@@ -372,3 +375,83 @@ int conf_files_list_dropins(
 
         return conf_files_list_strv(ret, ".conf", root, 0, (const char* const*) dropin_dirs);
 }
+
+/**
+ * Open and read a config file.
+ *
+ * The <fn> argument may be:
+ * - '-', meaning stdin.
+ * - a non-absolute path. In this case <config_dirs> are searched.
+ * - an absolute path. In this case <fn> is opened directly.
+ */
+int conf_file_read(
+                const char *root,
+                const char **config_dirs,
+                const char *fn,
+                parse_line_t parse_line,
+                void *userdata,
+                bool ignore_enoent,
+                bool *invalid_config) {
+
+        _cleanup_fclose_ FILE *_f = NULL;
+        _cleanup_free_ char *_fn = NULL;
+        unsigned v = 0;
+        FILE *f;
+        int r;
+
+        assert(fn);
+
+        if (streq(fn, "-")) {
+                f = stdin;
+                fn = "<stdin>";
+
+                log_debug("Reading config from stdin%s", special_glyph(SPECIAL_GLYPH_ELLIPSIS));
+
+        } else {
+                r = search_and_fopen(fn, "re", root, config_dirs, &_f, &_fn);
+                if (r == -ENOENT && ignore_enoent) {
+                        log_debug_errno(r, "Failed to open \"%s\", ignoring: %m", fn);
+                        return 0; /* No error, but nothing happened. */
+                }
+                if (r < 0)
+                        return log_error_errno(r, "Failed to read '%s': %m", fn);
+
+                f = _f;
+                fn = _fn;
+                log_debug("Reading config file \"%s\"%s", fn, special_glyph(SPECIAL_GLYPH_ELLIPSIS));
+        }
+
+        r = 1;  /* We entered the part where we may modify state. */
+
+        for (;;) {
+                _cleanup_free_ char *line = NULL;
+                bool invalid_line = false;
+                int k;
+
+                k = read_stripped_line(f, LONG_LINE_MAX, &line);
+                if (k < 0)
+                        return log_error_errno(k, "Failed to read '%s': %m", fn);
+                if (k == 0)
+                        break;
+
+                v++;
+
+                if (IN_SET(line[0], 0, '#'))
+                        continue;
+
+                k = parse_line(userdata, fn, v, line, &invalid_line);
+                if (k < 0 && invalid_line) {
+                        /* Allow reporting with a special code if the caller requested this. */
+                        if (invalid_config)
+                                *invalid_config = true;
+                } else
+                        /* The first error, if any, becomes our return value. */
+                        RET_GATHER(r, k);
+        }
+
+        if (ferror(f))
+                RET_GATHER(r, log_error_errno(errno_or_else(EIO),
```

Comment:
I don't get this. For `errno_or_else` to work, `errno` must be reset first. Plus, there's nothing that would set `errno` in the first place?